sql/src/main/java/io/crate/metadata/ReferenceToLiteralConverter.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.metadata;
+
+import io.crate.analyze.symbol.Literal;
+import io.crate.analyze.symbol.Reference;
+import io.crate.analyze.symbol.Symbol;
+import io.crate.types.DataType;
+import io.crate.types.ObjectType;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+
+import java.util.List;
+import java.util.Map;
+
+public class ReferenceToLiteralConverter extends ReplacingSymbolVisitor<ReferenceToLiteralConverter.Context> {
+
+    private static final ESLogger LOGGER = Loggers.getLogger(ReferenceToLiteralConverter.class);
+
+    public static class Context {
+        private final List<Reference> references;
+        private final Object[] values;
+
+        public Context(List<Reference> references, Object[] values) {
+            this.references = references;
+            this.values = values;
+        }
+
+        public Symbol resolveReferenceValue(Reference reference) {
+            for (int i = 0; i < references.size(); i++) {
```

Comment:
is this called per row? if so we should extract things which are known before the iteration